pkg/api/proxy.go

```diff
@@ -0,0 +1,86 @@
+package api
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/dchest/siphash"
+	"github.com/gorilla/mux"
+
+	"zotregistry.dev/zot/pkg/api/config"
+)
+
+func ClusterProxy(config *config.Config) func(http.HandlerFunc) http.HandlerFunc {
+	return func(next http.HandlerFunc) http.HandlerFunc {
+		return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
+			requestPath := request.URL.Path
+			fmt.Printf("Received path: %s\n", requestPath)
+
+			// if no cluster or single-node cluster, handle locally
+			if config.Cluster == nil || len(config.Cluster.Members) == 1 {
+				next.ServeHTTP(response, request)
+				return
+			}
+
+			vars := mux.Vars(request)
+			name, ok := vars["name"]
+
+			if !ok || name == "" {
+				response.WriteHeader(http.StatusNotFound)
+				return
+			}
+
+			h := siphash.New([]byte(config.Cluster.HashKey))
+			h.Write([]byte(name))
+			sum64 := h.Sum64()
+
+			targetMember := config.Cluster.Members[sum64%uint64(len(config.Cluster.Members))]
+			fmt.Printf("Target member is %v\n", targetMember)
+
+			// from the member list and our DNS/IP address, figure out if this request should be handled locally
+			localMember := fmt.Sprintf("%s:%s", config.HTTP.Address, config.HTTP.Port)
+			if targetMember == localMember {
+				fmt.Println("Member is local. Handling locally")
+				next.ServeHTTP(response, request)
+				return
+			}
+			fmt.Println("Target member is not local. Forwarding")
+
+			handleHttpForward(response, request, targetMember)
+		})
+	}
+}
+
+func handleHttpForward(w http.ResponseWriter, req *http.Request, targetMember string) {
+	fmt.Printf("incoming req: %v\n", *req)
+	fmt.Printf("Incoming URL is %v\n", req.URL)
+	url := req.URL
+	url.Scheme = "http"
```

Comment:
For some reason, it is required to manually set the Scheme here. Without this line, the request doesn't go through and there's an error about an `unsupported scheme ""` (since it is blank).

I don't have any explanation for why the original request did not have this populated already.

Reply:
yeah, they are not populated. you can use req.TLS to check if it's http or https.